### ISSUE TITLE
Unable to receive KafkaError using the channel producer

### ISSUE THREAD

**Issue by sivann** [confluentinc/confluent-kafka-go#200]
Description
===========
I used the channel producer from the example. When brokers are down I get the errors from the library:

3|1528270208.663|FAIL|rdkafka#producer-1| [thrd:pkgtest.xxx.com:9092/bootstrap]: pkgtest.inaccess.com:9092/bootstrap: Connect to ipv4#xxx.xxx.5.40:9092 failed: Connection refused
%3|1528270208.663|ERROR|rdkafka#producer-1| [thrd:pkgtest.xxx.com:9092/bootstrap]: pkgtest.inaccess.com:9092/bootstrap: Connect to ipv4#xxx.xxx.5.40:9092 failed: Connection refused
%3|1528270208.663|ERROR|rdkafka#producer-1| [thrd:pkgtest.xxx.com:9092/bootstrap]: 1/1 brokers are down

But I can't get any generic errors like ALL_BROKERS_DOWN in the producer.Events(). I presume from the documentation that those are KafkaError events.

How to reproduce
================
Run the channel producer example with bogus brokers. I would expect it to say something like "Ignored event: all brokers down".

Checklist
=========
Please provide the following information:

 - [x] librdkafka: 0.11.4
 - [x] confluent-kafka-go: 1112e2c08a15bce669a99cc35284a690fd2086b8
 - [x] Operating system: MacOSX Sierra

**Comment by edenhill** [comment]
This will be fixed in the upcoming v0.11.5 release.

**Comment by sivann** [comment]
@edenhill if you fix this, it would be also preferable to allow the current errors output to be redirectable to a file descriptor instead of the hardcoded stderr. Thanks.

**Comment by edenhill** [comment]
Those are separate things, error events and logs.
However, if no error event handler is registered (which was the case before this fix) then all errors will be logged, so with this fix in there will be less logging.

As for logs, you are right, we should expose them as Log event types in the API. :+1:

**Comment by edenhill** [comment]
Tracked in #54

**Comment by vdepatla** [comment]
@edenhill  when is the v0.11.5 release ?

**Comment by edenhill** [comment]
Within 5 days

**Comment by devrimbaris** [comment]
@edenhill I am testing with v0.11.5 release with my producer and it seems I still can't intercept the error if the broker is down.

**Comment by rnpridgeon** [comment]
@devrimbaris , I will try to reproduce later this week and report back.

**Comment by edenhill** [comment]
The v0.11.5 release of the Go client was delayed due to the Admin API PR, we're looking to release within a couple of weeks.

But as long as you're using librdkafka v0.11.5 you can use the master branch version of the Go client which should support the new generic Error events.

And yes, you should be able to get broker errors like that as Error events now.

**Comment by onorua** [comment]
I believe it is implemented in v0.11.6:
```
	for e := range p.Producer.Events() {
		switch ev := e.(type) {
		case *kafka.Message:
                        // reply message ....
		case kafka.Error:
			if ev.Code() == kafka.ErrAllBrokersDown {
				// do what is required when brokers are down
			}
		}
```
I believe this issue can be closed

**Comment by pragmaticpi** [comment]
If the kafka broker went offline and the message is not produced to the channel, then where I will get the delivery report? I am not getting delivery failed in either `message.TopicPartition.Error` or `kafka.Error`.

**Comment by edenhill** [comment]
@pragmaticpi Are you using the channel producer? If so it will always block and eventually accept the message.
Any messages accepted by Produce() or the produce channel will eventually be successfully delivered, fail with a permanent error, or fail by timeout (message.timeout.ms), whichever comes first.
All those results are propagated through the delivery report as a *Message event on the Events channel.

**Comment by pragmaticpi** [comment]
Yeah, I am using channel producer using [this](https://github.com/confluentinc/confluent-kafka-go/blob/master/examples/producer_channel_example/producer_channel_example.go) example. 
According to docs, `Delivery reports are by default emitted on the .Events() channel as *kafka.Message and you should check msg.TopicPartition.Error for nil to find out if the message was succesfully delivered or not`. 
So basically, I have to wait till `message.timout.ms` to know whether message delivered or not..right? Is there a way to configure this timeout property?
Also, what types of error will come into `TopicPartiton.Error`? 

Thanks

**Comment by edenhill** [comment]
Just specify `"message.timeout.ms": 12345` in the ConfigMap you pass to NewProducer().

There is no definitive list of errors given in TopicPartition.Error, but we're working on it.
But from the application points of view it should simply consider the message as failed, the application should not retry - it is better to configure the producer to perform the retries since it has the logic to know how, when and why.
The default producer settings should be good.

**Comment by pragmaticpi** [comment]
Thanks for the answer. I configured `message.timeout.ms` to 2000ms and after deliberately stopping the broker and I am able to get delivery failed report through `topicPartition.Error`.

Right now I got data streaming at every 5 sec or so. And if the broker goes down and I got failed report for my last message, should I let the other messages be produced to the channel and wait for the broker to get online? Is the timeout configuration helpful in this case? Or is there any other way to approach this?

Thanks

**Comment by edenhill** [comment]
Instead of trying to keep track on the cluster/broker state (which is hard and multi-faceted, and something the client already does for you),  use `message.timeout.ms` as your data policy;  how long does it make sense to try to produce this type of data? And then leave it up to the client to do its best to produce the data within that constraint.
